keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,843 @@
+identified as those that are in the `system-cluster-critical` or
+`system-node-critical` [priority
+classes](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical).
+
+Upon shutdown Kubelet will:
+
+1. Gracefully terminate all non critical system pods with a gracePeriodOverride
+computed as `min(podSpec.terminationGracePeriodSeconds, ShutdownGracePeriod)`
+2. Gracefully terminate all critical system pods with gracePeriodOverride of 2
+seconds
+
+Kubelet will use the same existing
+[killPod](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/pod_workers.go#L292)
+function to perform the termination of pods, using `gracePeriodOverride` to set
+the appropriate grace period. During the termination process, normal [pod
+termination](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)
+processes will apply, e.g. preStopHooks will be called,  SIGTERM to containers
+delivered, etc.
+
+2 seconds as gracePeriodOverride for critical system pods was decided to ensure
+that they can also perform a graceful shutdown and 2 seconds is currently
+[defined](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/kuberuntime/kuberuntime_container.go#L626-L629)
+as the minimum grace period defined in the kubelet.
+
+POC: I’ve prototyped an initial POC
+[here](https://github.com/bobbypage/kubernetes/tree/shutdown) of the proposed
+implementation on the `shutdown` branch.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+*   Kubelet does not receive shutdown event or is able to create inhibitor lock
+    *   Mitigation: Kubelet does not provide graceful shutdown to pods (same as
+        today’s existing behavior)
+*   Kublet is unable to update `InhibitDelayMaxSec` in logind to match that of
+    `kubeletConfig.ShutdownGracePeriod`
+    *   If there are multiple logind configuration file overrides in
+        `/etc/systemd/logind.conf.d/`, logind will use the config file with the
+        lexicographically latest name. As a result in rare cases, the kubelet’s
+        InhibitDelayMaxSec conf file override may be overwritten by another
+        config file (possibly placed by another service on the machine).
+    *   Mitigation: Kubelet will use current value of `InhibitDelayMaxSec` from
```

Comment:
Shouldn't we then name it for example `00-kubelet.conf`? IOW, if the kubelet is expected to manage pod/machine shutdown, then it seems reasonable for the kubelet to (best-effort) take precendence.

Reply:
That's a good point -- the name of the file can be set to ensure kubelet has higher chance of being selected. I'll take a closer look at the best naming here so the config has a higher chance of being selected.